chew.go

```diff
@@ -41,17 +77,122 @@ var contentTypeProcessors = map[string]func(io.Reader, string) ([]common.Chunk,
 	contentTypeEPUB:     document.ProcessEpub,
 }
 
+type Chew struct {
+	config       common.Config
+	httpClient   *http.Client
+	rateLimiter  RateLimiter
+	robotsCache  map[string]*robotstxt.RobotsData
+	robotsMu     sync.RWMutex
+	lastAccess   map[string]time.Time
+	lastAccessMu sync.Mutex
+	proxyIndex   int
+	proxyMu      sync.Mutex
+}
+
+type RateLimiter interface {
+	Wait(context.Context) error
+}
+
+func (c *Chew) SetRateLimiter(rl RateLimiter) {
+	c.rateLimiter = rl
+}
+
 /*
-Transcribe uses the Google Cloud Speech-to-Text API to transcribe an audio file. It takes
-a context, the filename of the audio file to transcribe, and a TranscribeOptions struct which
-contains the Google Cloud credentials, the GCS bucket to upload the audio file to, and the language code
-to use for transcription. It returns the transcript of the audio file as a string and an error if the
-transcription fails.
+NewConfig allows you to set the configuration options for URL processing. It takes a Config struct.
+
+Usage:
+
+	config := chew.Config{
+		UserAgent:       "MyBot/1.0 (+https://example.com/bot)",
+		RetryLimit:      3,
+		RetryDelay:      5 * time.Second,
+		CrawlDelay:      10 * time.Second,
+		ProxyList:       []string{"http://proxy1.com", "http://proxy2.com"},
+		RateLimit:       2 * time.Second,
+		RateBurst:       3,
+		IgnoreRobotsTxt: false,
+	}
+
+	chew.NewConfig(config)
+*/
+func New(config common.Config) *Chew {
+	c := &Chew{
+		config:      config,
+		robotsCache: make(map[string]*robotstxt.RobotsData),
+		lastAccess:  make(map[string]time.Time),
+	}
+	c.initHTTPClient()
+
+	limit := rate.Every(config.RateLimit)
+	c.rateLimiter = rate.NewLimiter(limit, config.RateBurst)
+
+	return c
+}
+
+/*
+Transcribe is a function that transcribes audio files using either the Google Cloud Speech-to-Text API
+or the Whisper API. It handles uploading the audio file to Google Cloud Storage if necessary,
+manages the transcription process, and returns the resulting transcript.
+
+For detailed usage instructions, see the TranscribeOptions struct documentation.
 */
 var Transcribe = transcribe.Transcribe
 
+/*
+The TranscribeOptions struct contains the options for transcribing an audio file. It allows the user
+to specify the Google Cloud credentials, the GCS bucket to upload the audio file to, the language code
+to use for transcription, a potion to enable diarization (the process of separating and labeling
+speakers in an audio stream) including the min and max speakers and
+an option to clean up the audio file from GCS after transcription is complete.
```

Comment:
Please write something like this


```suggestion
an option to clean up the audio file from Google Cloud Speech-to-Text (GCS) after transcription is complete.
```

Reply:
done